dev/scripts/src/alluxio.org/build/cmd/docker.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	// parse available docker images in docker.yml
+	{
+		wd, err := os.Getwd()
+		if err != nil {
+			return stacktrace.Propagate(err, "error getting current working directory")
+		}
+		dockerYmlPath := filepath.Join(wd, opts.dockerYmlFile)
+		content, err := ioutil.ReadFile(dockerYmlPath)
+		if err != nil {
+			return stacktrace.Propagate(err, "error reading file at %v", dockerYmlPath)
+		}
+		if err := yaml.Unmarshal(content, &opts.dockerImages); err != nil {
+			return stacktrace.Propagate(err, "error unmarshalling docker images from:\n%v", string(content))
+		}
+	}
+
+	image, ok := opts.dockerImages[opts.image]
+	if !ok {
+		return stacktrace.NewError("must provide valid 'image' arg")
+	}
+	alluxioVersion, err := alluxioVersionFromPom()
+	if err != nil {
+		return stacktrace.Propagate(err, "error parsing version string")
+	}
+	opts.buildOpts = tOpts
+	if opts.artifactOutput != "" {
+		artifact, err := artifact.NewArtifact(
+			artifact.DockerArtifact,
+			opts.outputDir,
+			strings.ReplaceAll(opts.dockerImages[opts.image].TargetName, versionPlaceholder, alluxioVersion),
+			alluxioVersion,
+			map[string]string{"image": opts.image})
+		if err != nil {
+			return stacktrace.Propagate(err, "error adding artifact")
+		}
+		return artifact.WriteToFile(opts.artifactOutput)
+	}
+
+	if opts.tarballPath == "" {
+		if err := buildTarball(tOpts); err != nil {
+			return stacktrace.Propagate(err, "error building tarball")
+		}
+		opts.tarballPath = filepath.Join(opts.outputDir,
+			strings.ReplaceAll(opts.targetName, versionPlaceholder, alluxioVersion))
+	}
+
+	// docker logic
+	if err != nil {
+		return stacktrace.Propagate(err, "error finding repo root")
+	}
+	dockerWs := filepath.Join(findRepoRoot(), "integration", "docker")
+	if err := command.RunF("cp %v %v", opts.tarballPath, filepath.Join(dockerWs, tempAlluxioTarballName)); err != nil {
+		return stacktrace.Propagate(err, "error copying tarball to docker workspace")
+	}
+	defer os.RemoveAll(filepath.Join(dockerWs, tempAlluxioTarballName))
+
+	if err := image.build(opts, alluxioVersion, true); err != nil {
+		return stacktrace.Propagate(err, "error building image %v", opts.image)
+	}
+
+	return nil
+}
+
+func (i *DockerImage) build(opts *dockerBuildOpts, alluxioVersion string, save bool) error {
+	dockerWs := filepath.Join(findRepoRoot(), i.BuildDir)
+	if i.Dependency != "" {
+		dep, ok := opts.dockerImages[i.Dependency]
+		if !ok {
+			return stacktrace.NewError("%v not found in list of docker images", i.Dependency)
+		}
+		if err := dep.build(opts, alluxioVersion, false); err != nil {
+			return stacktrace.Propagate(err, "error building dep %v", i.Dependency)
+		}
+	}
+	i.Tag = strings.ReplaceAll(i.Tag, versionPlaceholder, alluxioVersion)
+	i.TargetName = strings.ReplaceAll(i.TargetName, versionPlaceholder, alluxioVersion)
+	i.outputTarball = fmt.Sprintf("%v/%v",
+		opts.outputDir, strings.ReplaceAll(i.TargetName, versionPlaceholder, alluxioVersion))
```

Comment:
maybe create a `i.ReplaceVersionPlacholders(alluxioVersion)` to cover all of the various fields that need replacement? would be called immediately after retrieving the image from the map, could also be called `Init()`.

Reply:
updated